wave_lang/kernel/wave/water.py

```diff
@@ -446,3 +450,65 @@ def add_opt(pipeline):
 
     with module.context:
         return Module.parse(result)
+
+
+def apply_water_lowering_passes(mlir_text: str) -> str:
+    """Apply Water Wave lowering passes using subprocess water-opt.
+
+    This function applies the standard Wave lowering passes:
+    - water-wave-detect-normal-forms
+    - water-wave-propagate-elements-per-thread
+    - lower-wave-to-mlir
+    - canonicalize
+    - cse
+
+    Args:
+        mlir_text: Input Wave dialect MLIR as string
+
+    Returns:
+        Optimized MLIR as string after applying the passes
+
+    Raises:
+        RuntimeError: If water-opt is not available or passes fail
+    """
+    binary = get_water_opt()
+
+    # Define the pass pipeline for Wave lowering
+    pipeline = [
+        "water-wave-detect-normal-forms",
+        "water-wave-propagate-elements-per-thread",
+        "lower-wave-to-mlir",
+        "canonicalize",
+        "cse",
+    ]
+
+    try:
+        result = subprocess.check_output(
+            [
+                binary,
+                make_linear_pass_pipeline(pipeline),
+            ],
+            input=mlir_text,
+            text=True,
+        )
+
+        # Clean Wave-specific attributes that cause issues in later pipelines
+        # TODO(tyb): do this in the very last pass of water pipeline instead
```

Comment:
Open an issue for this so we can track it, I really dislike having random regular expressions mutating the IR. This is difficult to maintain and will be easily forgotten here.

Even better if you just vibe code the pass that removes the normal form tag, it shouldn't be more than a couple dozen lines scattered across files.

Reply:
See #595